keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md

```diff
@@ -195,6 +242,176 @@ import (
+
+In all other respects, the expected behavior of the cloud controller managers is not changing,
+and as such the original tests for in-tree cloud controllers continue to be relevant and
+necessary.
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+#### Prerequisite testing update
+
+The behavior of the cloud controllers is not changing with respect to cluster functioning,
+as such the prerequisite for this KEP is that all the current cloud controller related
+tests are passing. When the external CCMs are enabled, these tests should continue to pass.
+
+As described in the [non-goals](#non-goals) section, testing of individual cloud provider CCMs
+is not in scope for this KEP. Each provider is expected to own the tests related to their
+specific cloud platform.
+
+#### Unit tests
+
+This KEP describes a process for extracting and migrating current in-tree code
+to external repositories. The notion of unit style testing takes on a different
+meaning in this perspective as the focus for the unit testing will move to those
+external repositories.
+
+SIG Cloud Provider maintains a [framework for creating CCMs][ccmframework] which
+contains unit testing for the core controller loops of a CCM implementation. See:
+
+- https://github.com/kubernetes/cloud-provider/blob/master/controllers/node/node_controller_test.go
+- https://github.com/kubernetes/cloud-provider/blob/master/controllers/nodelifecycle/node_lifecycle_controller_test.go
+- https://github.com/kubernetes/cloud-provider/blob/master/controllers/route/route_controller_test.go
+- https://github.com/kubernetes/cloud-provider/blob/master/controllers/service/controller_test.go
+
+Cloud providers who create external CCMs will be responsible for providing unit
+testing within their own repositories.
+
+#### Integration tests
+
+Integration testing on this change is complicated by the fact that any testing
+with concrete CCM implementations requires cloud infrastructure from the same
+provider. While it would be possible to create a mock provider for these type
+of scenarios, the SIG considers this testing to be a low value when compared to
+the e2e test which are running on concrete providers.
+
+For the reasons stated above, this KEP focuses in e2e testing over integration.
+
+#### e2e tests
+
+This KEP will leverage the existing e2e test suite for the majority of testing.
+Given that this KEP proposes no behavioral changes to the functioning the of
+the cloud controllers, the existing tests will provide a valuable signal to
+ensure that nothing has been broken during the migration from in-tree to
+external CCM.
+
+The following pull request is the first in a series (documented in the request)
+which enable external CCMs by default for all GCE/GCP testing from the Kubernetes
+repository:
+
+- https://github.com/kubernetes/kubernetes/pull/117503
```

Comment:
Hmm - why did we merge https://github.com/kubernetes/kubernetes/pull/117503
without having the KEP approved for beta?

@dims 

[Depending on what I will see on the PRR itself, I would say a possible outcome might be a need to revert that change... ?]

Reply:
In @dims defense, we wanted the CI signal in really early in the release so we can establish a stable CI signal with all the in-tree cloud providers disabled before any other feature work started in 1.29. Agreed we can revert if all the PRR concerns aren't addressed, the good thing now is that we can revert the feature gate flip without having to revert all the CI jobs to use in-tree providers again.